### PR TITLE
Use fixed c# statement when accessing fixed struct members

### PR DESCRIPTION
**Pull request by giladlevi**

### CODE REVIEW

File: src/Generator/Generators/CSharp/CSharpMarshal.cs [hunk 1]
```diff
@@ -87,9 +87,12 @@ public override bool VisitArrayType(ArrayType array, TypeQualifiers quals)
             {
                 case ArrayType.ArraySize.Constant:
                     var supportBefore = Context.SupportBefore;
-                    string value = Generator.GeneratedIdentifier("value");
+                    var value = Generator.GeneratedIdentifier("value");
                     supportBefore.WriteLine("{0}[] {1} = null;", array.Type, value, array.Size);
-                    supportBefore.WriteLine("if ({0} != null)", Context.ReturnVarName);
+                    var fixedPtrName = $"{array.Type}Ptr";
+                    supportBefore.WriteLine("fixed ({0}* {1} = {2})", array.Type, fixedPtrName, Context.ReturnVarName);
```
Comment (ddobrev) [inline src/Generator/Generators/CSharp/CSharpMarshal.cs]: Please use an interpolated string here ($""), as in the line above.

File: src/Generator/Generators/CSharp/CSharpMarshal.cs [hunk 2]
```diff
@@ -87,9 +87,12 @@ public override bool VisitArrayType(ArrayType array, TypeQualifiers quals)
...
+                    supportBefore.WriteStartBraceIndent();
+                    supportBefore.WriteLine("if ({0} != null)", fixedPtrName);
```
Comment (ddobrev) [inline src/Generator/Generators/CSharp/CSharpMarshal.cs]: Interpolate the string.

File: src/Generator/Generators/CSharp/CSharpMarshal.cs [hunk 3]
```diff
@@ -434,6 +438,10 @@ public override bool VisitArrayType(ArrayType array, TypeQualifiers quals)
                             supportBefore.WriteLine("if (value.Length != {0})", array.Size);
                             ThrowArgumentOutOfRangeException();
                         }
+                        var fixedPtrName = $"{array.Type}Ptr";
+                        supportBefore.WriteLine("fixed ({0}* {1} = {2})", array.Type, fixedPtrName, Context.ReturnVarName);
```
Comment (ddobrev) [inline src/Generator/Generators/CSharp/CSharpMarshal.cs]: Interpolate the string.

**Comment by tritao** [comment]
Hi @giladlevi, thanks for the PR.

Do you mind adding a test for this failure?

You can add it to https://github.com/mono/CppSharp/blob/master/tests/Common/Common.h#L636 and https://github.com/mono/CppSharp/blob/master/tests/Common/Common.Tests.cs#L541.

**Comment by ddobrev** [comment]
@giladlevi thank you for your work. I have to ask you, however, for some additional adjustment. First of all, please add a test, you should use CSharp.h. Second, ensure all tests pass locally for you before submitting your pull request because you can see at https://ci.appveyor.com/project/tritao/cppsharp/build/1.0.1971 that your changes introduce other errors.

**Comment by tritao** [comment]
Hey @giladlevi, any update on the PR?

**Comment by giladlevi** [comment]
Hey @tritao i pushed force and override the first commit now it: 5302f5870cafcade6b1d81d34ff19e4b131cd567
+
I added support for parsing macros such **#define MY_MACRO '1'** in ParseMacroExpression.
I need it for generates enums with asci codes.

**Comment by ddobrev** [comment]
@giladlevi thank you very much for your work. I would just like to give you a minor recommendation:

    var arrayType = field.Type as ArrayType ?? field.QualifiedType.Type.Desugar() as ArrayType;

can be shortened to:

    var arrayType = field.Type.Desugar() as ArrayType;

But that's not crucial so your choice. What's more important, thought, is that your second feature about the macro has no test. May I ask you to add one?

File: tests/CSharp/CSharp.cs [hunk 4]
```diff
@@ -35,6 +37,10 @@ public override void Preprocess(Driver driver, ASTContext ctx)
             ctx.SetClassAsValueType("QRect");
             ctx.SetClassAsValueType("StructTestArrayTypeFromTypedef");
             ctx.IgnoreClassWithName("IgnoredTypeInheritingNonIgnoredWithNoEmptyCtor");
+
+            var macroRegex = new Regex(@"(MY_MACRO_TEST_.*)");
+            var enumTest = ctx.GenerateEnumFromMacros("MyMacroTestEnum", (from unit in ctx.TranslationUnits where unit.FilePath == "<invalid>" || unit.FileName == "CSharp.h" from macro in unit.PreprocessedEntities.OfType<MacroDefinition>() let match = macroRegex.Match(macro.Name) where match.Success select macro.Name).ToArray());
```
Comment (ddobrev) [inline tests/CSharp/CSharp.cs]: Please move the LINQ to a new line and split it into more lines, one per statement.

File: tests/CSharp/CSharp.cs [hunk 5]
```diff
@@ -35,6 +37,10 @@ public override void Preprocess(Driver driver, ASTContext ctx)
...
+            enumTest.Namespace = new Namespace() {Name = "MacroTest"};
```
Comment (ddobrev) [inline tests/CSharp/CSharp.cs]: Please use this enum in the CSharp.Tests.cs, this way we'll know it's actually generated.

**Comment by ddobrev** [comment]
@giladlevi you don't need to open a new pull request each time. Just amend your top commit and then push force, your pull will be automatically updated.

**Comment by ddobrev** [comment]
Also, you might be missing a using:

https://ci.appveyor.com/project/tritao/cppsharp/build/1.0.2012#L336

**Comment by ddobrev** [comment]
@giladlevi thank you very much for your work.

**Pull request merged**